libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/MapRouteLine.java

```diff
@@ -0,0 +1,513 @@
+    Feature feature = Feature.fromGeometry(Point.fromLngLat(
+      leg.steps().get(index).maneuver().location().longitude(),
+      leg.steps().get(index).maneuver().location().latitude()
+    ));
+    feature.addStringProperty("waypoint", index == 0 ? "origin" : "destination");
+    return feature;
+  }
+
+  private void updateRoutesForNewIndex() {
+    for (int i = 0; i < routeFeatureCollections.size(); i++) {
+      String shieldLayerId = String.format(
+        Locale.US, RouteConstants.ID_FORMAT, ROUTE_SHIELD_LAYER_ID, i
+      );
+      updatePrimaryShieldRoute(shieldLayerId, i);
+
+      String routeLayerId = String.format(
+        Locale.US, RouteConstants.ID_FORMAT, ROUTE_LAYER_ID, i
+      );
+      updatePrimaryRoute(routeLayerId, i);
+    }
+  }
+
+  private void updatePrimaryRoute(String layerId, int index) {
+    Layer layer = mapboxMap.getLayer(layerId);
+    if (layer != null) {
+      layer.setProperties(
+        PropertyFactory.lineColor(match(
+          Expression.toString(get(RouteConstants.CONGESTION_KEY)),
+          color(index == primaryRouteIndex ? routeDefaultColor : alternativeRouteDefaultColor),
+          stop("moderate", color(index == primaryRouteIndex ? routeModerateColor : alternativeRouteModerateColor)),
+          stop("heavy", color(index == primaryRouteIndex ? routeSevereColor : alternativeRouteSevereColor)),
+          stop("severe", color(index == primaryRouteIndex ? routeSevereColor : alternativeRouteSevereColor))
+          )
+        )
+      );
+      if (index == primaryRouteIndex) {
+        mapboxMap.removeLayer(layer);
+        mapboxMap.addLayerBelow(layer, WAYPOINT_LAYER_ID);
+      }
+    }
+  }
+
+  private void updatePrimaryShieldRoute(String layerId, int index) {
+    Layer layer = mapboxMap.getLayer(layerId);
+    if (layer != null) {
+      layer.setProperties(
+        PropertyFactory.lineColor(index == primaryRouteIndex ? routeShieldColor : alternativeRouteShieldColor)
+      );
+      if (index == primaryRouteIndex) {
+        mapboxMap.removeLayer(layer);
+        mapboxMap.addLayerBelow(layer, WAYPOINT_LAYER_ID);
+      }
+    }
+  }
+
+  private void findRouteBelowLayerId() {
+    if (belowLayer == null || belowLayer.isEmpty()) {
+      List<Layer> styleLayers = mapboxMap.getLayers();
+      for (int i = 0; i < styleLayers.size(); i++) {
+        if (!(styleLayers.get(i) instanceof SymbolLayer)
+          // Avoid placing the route on top of the user location layer
+          && !styleLayers.get(i).getId().contains(RouteConstants.MAPBOX_LOCATION_ID)) {
+          belowLayer = styleLayers.get(i).getId();
+        }
+      }
+    }
+  }
+
+  private void obtainStyledAttributes(Context context) {
+    TypedArray typedArray = context.obtainStyledAttributes(styleRes, R.styleable.NavigationMapRoute);
+
+    // Primary Route attributes
+    routeDefaultColor = typedArray.getColor(R.styleable.NavigationMapRoute_routeColor,
```

Comment:
@LukasPaczos this should be returning a `@ColorInt` that can be used with the expression right?